golang-github-osbuild-composer.spec

```diff
@@ -68,16 +69,30 @@ export GOFLAGS=-mod=vendor
 %gobuild -o _bin/osbuild-worker %{goipath}/cmd/osbuild-worker
 %gobuild -o _bin/osbuild-image-tests %{goipath}/cmd/osbuild-image-tests
 
-# Build test binaries with `go test -c`, so that they can take advantage of
-# golang's testing package. The golang rpm macros don't support building them
-# directly. Thus, do it manually, taking care to also include a build id.
-
-TEST_LDFLAGS="${LDFLAGS:-} -B 0x$(od -N 20 -An -tx1 -w100 /dev/urandom | tr -d ' ')"
-
-go test -c -tags=integration -ldflags="${TEST_LDFLAGS}" -o _bin/osbuild-tests %{goipath}/cmd/osbuild-tests
-go test -c -tags=integration -ldflags="${TEST_LDFLAGS}" -o _bin/osbuild-dnf-json-tests %{goipath}/cmd/osbuild-dnf-json-tests
-go test -c -tags=integration -ldflags="${TEST_LDFLAGS}" -o _bin/osbuild-weldr-tests %{goipath}/internal/weldrcheck/
-go test -c -tags=integration -ldflags="${TEST_LDFLAGS}" -o _bin/osbuild-rcm-tests %{goipath}/cmd/osbuild-rcm-tests
+# The tests need to be compiled by `go test -c`.
+# Go rpm macros have gotest, but it can be used only in check to run
+# unit tests during rpmbuild.
+#
+# `go test -c` has very similar to arguments to `go build`. The following
+# macro uses that fact. It takes gobuild macro and substitutes `go build`
+# with `go test -c` inside the macro. This way we get working `go test -c`
+# command which should be in sync with changes in gobuild macro.
+#
+# Because gobuild macro gets expanded to a string containing apostrophes
+# and shell expansions (which we don't want to expand just yet), heredoc
+# is used to pipe the macro into sed. The apostrophes around HEREDOC
+# put it into non-expansion mode.
+%define gobuildtest(o:) %(sed "s/go build/go test -c/" <<'HEREDOC'
```

Comment:
Oh, this is so awesome!

Reply:
**very** impressive hack, but it is quite a hack. For example, we're passing command line options to `go test` that are documented for `go build` only.

But wow, it is quite something :smiley:

Reply:
From `go help test`:

> In addition to the build flags, the flags handled by 'go test' itself are:
> ...

Reply:
Also, from the first line of `go help test`:

> usage: go test [build/test flags] [packages] [build/test flags & test binary flags]

Reply:
This is also how I figured out the flags in my original attempts at this.

Reply:
Ahhh I overread that. Thanks!